### PR TITLE
library of interfaces to fetch system resource information

### PR DESCRIPTION
**Pull request by rahuyada**
1: System resource Information related to  memory, CPU, stat, networking,
security etc.
2: Currently most of such information is read from /proc and /sys.
3: Add basic infrastructure and code for some of memory and network related
information.
4: Update README file with the use cases and design details.
5: Added example codes in README.

Signed-off-by: Rahul Yadav <rahul.x.yadav@oracle.com>

### CODE REVIEW

**Comment by rahuyada** [comment]
@hallyn @brauner I am not able to add reviewers. Please add yourself as reviewers and review the code.

**Comment by brauner** [comment]
Hi Rahul,

Funny, I just mentioned your patchset to someone today. :)


On Wed, Mar 28, 2018, 14:47 rahuyada <notifications@github.com> wrote:

> @hallyn <https://github.com/hallyn> @brauner <https://github.com/brauner>
> I am not able to add reviewers. Please add yourself as reviewers and review
> the code.
>
I just added @hallyn and myself. Thanks for the new PR. Good that it worked
out.

Thanks!
Christian

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+%.o: %.c $(DEPS)
+	$(CC) -c -o $@ $< $(CFLAGS)
+
+all: $(OBJ)
+	$(CC) -shared -o $(LIB) $^ $(CFLAGS)
+
+.PHONY : clean
+clean:
+	$(RM) $(LIB) $(OBJ) $(TEST)
```
Comment (brauner) [inline Makefile]: I guess that'll do for now but we'll need a proper build system such as autotools or meson. Whatever works.

File: resnet.c [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+		*name++ = *p++;
+	}
+	*name++ = '\0';
+	return p;
+}
+
+/* Scan network interface stat line for interface and populate all relevant
+ * information.
+ */
+static inline int scan_net_stat(char *buf, res_net_ifstat_t *i, int ver)
```
Comment (brauner) [inline resnet.c]: Hm, that is a lot of parsing. Any reason why you wouldn't be going through `netlink`?

File: resource.c [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+		return getmeminfo(res_id, out, hint, pid, flags);
+
+	/* Check if net proc file is needed to open */
+	if (res_id >= NET_MIN && res_id < NET_MAX)
+		return getnetinfo(res_id, out, hint, pid, flags);
+
+	switch (res_id) {
+	case RES_KERN_RELEASE:
+		uname(&t);
+		strcpy(out, t.release);
```
Comment (brauner) [inline resource.c]: You don't know the size of out. That looks like a potential buffer overflow. :)

File: resource.h [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+ * returned.
+ */
+#define RES_UNIT_OUT_SIZE	256
+
+/* This union is used to return resource information of various types */
+union r_data {
+	int i;
+	size_t sz;
+	long l;
+	char str[RES_UNIT_OUT_SIZE];
```
Comment (brauner) [inline resource.h]: What exactly is this for?

Comment (brauner) [reply resource.h]: The union itself is pretty opaque right now. Can you expand what the individual fields are for, please?

File: resource.h [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+	void *ptr;
+};
+
+/* In case of res_read_blk, each resource information will be represented by
+ * following structure.
+ */
+typedef struct res_unit {
+	int status;
+	unsigned int res_id;
+	void *hint;
```
Comment (brauner) [inline resource.h]: `hint` is used to pass what, e.g. the name of the interface to query?

File: resource.c [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+	if (res_id >= MEM_MIN && res_id < MEM_MAX)
+		return getmeminfo(res_id, out, hint, pid, flags);
+
+	/* Check if net proc file is needed to open */
+	if (res_id >= NET_MIN && res_id < NET_MAX)
+		return getnetinfo(res_id, out, hint, pid, flags);
+
+	switch (res_id) {
+	case RES_KERN_RELEASE:
+		uname(&t);
```
Comment (brauner) [inline resource.c]: Seems to be missing error handling.

File: Makefile (hunk 1 above)
Comment (rahuyada) [reply Makefile]: I agree. I have started workng on such build system using autotools and can push it with next set of resource ids after this goes in.

File: resnet.c (hunk 2 above)
Comment (rahuyada) [reply resnet.c]: I took inspiration from netstat code when I was working on it, however I agree that I can add more comments.
Regarding netlink, you mean to fetch both stat and interface name using netlink right ? I remeber I tried that too but did not see any performance improvement so just did what netstat was doing. However if netlink is the way to go in future I guess I should change this. Let me know if I need to change this to netlink.

File: resource.c (hunk 3 above)
Comment (rahuyada) [reply resource.c]: will fix it.

File: resource.h (hunk 4 above)
Comment (rahuyada) [reply resource.h]: do you mean to put comments and explain what resource IDs will use what fields ?
This union is to provide user a way to get info without having to typecase return info in userland and was one of the suggestions on my draft. one field will can be used for multiple resoure IDs, for example both no of CPUs and no of Interfaces can be returned as integar.

File: resource.h (hunk 5 above)
Comment (rahuyada) [reply resource.h]: Yes name of interface is one such example. Basically hint is used to pass any additional info to the interface.

File: resource.c (hunk 6 above)
Comment (rahuyada) [reply resource.c]: Will fix it.

**Comment by rahuyada** [comment]
@brauner I have replied to the questions and have fixed couple of things according to comments. Please let me know how do you suggest to address other things, if any.

**Comment by rahuyada** [comment]
Hi Chris

Please have a look at my comments and my fixes according to your comments and let me know if there are things which I need to addres..

@hallyn Please have a look at the code changes and let me know your thoughts. 

Thanks
Rahul

File: README.md [hunk 7]
```diff
@@ -1,14 +1,167 @@
...
+    } res_blk_t;
+
+res_blk_t strcut is used in all res_*_blk interfaces.
+
+**3.2 functions**
+
+`res_blk_t *res_build_blk(int *res_ids, int res_count);`
+
+It allocates memory for resources and initiates them properly. res_ids
+holds and array of valid resource ids and res_count holds number of
```
Comment (hallyn) [inline README.md]: s/and/an

File: README.md [hunk 8]
```diff
@@ -1,14 +1,167 @@
...
+resource ids. It also initiates struct fields properly.
```
Comment (hallyn) [inline README.md]: s/initiates/initializes?

File: resnet.c [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+		/* If number of interfaces is not 0 then read interfaces
+		 * stat sequentially and put it in memory allocated
+		 * by user in out. Memory might be allocated for more interfaces
+		 * than there are on system. So read sequentially till either
+		 * all interfaces are read or memory allocated is totally
+		 * filled.
+		 */
+		if (out == NULL) {
+			errno = EINVAL;
+			return -1;
```
Comment (hallyn) [inline resnet.c]: leaking open fp?

File: resnet.c [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+		i = 0;
+
+		/* Memory is allocated for NET_ALLIFSTAT_SZ interfaces
+		 * initially and then reallocated in chunks of
+		 * NET_ALLIFSTAT_SZ till all information is read.
+		 * This allocated memory should be freed by consumer
+		 * of libresource interfaces.
+		 */
+		msz = sizeof(res_net_ifstat_t) * NET_ALLIFSTAT_SZ;
+		lst = (res_net_ifstat_t *)malloc(msz);
```
Comment (hallyn) [inline resnet.c]: Check for alloc failures.

File: resnet.c [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+
+		while (fgets(buf, sizeof(buf), fp) != NULL) {
+			if (i == NET_ALLIFSTAT_SZ) {
+				msz = (sizeof(res_net_ifstat_t) * inum) +
+					NET_ALLIFSTAT_SZ;
+				lst = (res_net_ifstat_t *)realloc(lst, msz);
```
Comment (hallyn) [inline resnet.c]: Proper check for (and recovery from) realloc failure?

File: resource.c [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+	if (!res_ids || res_count <= 0) {
+		errno = EINVAL;
+		return NULL;
+	}
+
+	/* Allocate memory to hold addresses of individual resource
+	 * information.
+	 */
+	res = (res_blk_t *)
+		malloc(sizeof(res_blk_t) + (sizeof(res_unit_t *)) * res_count);
```
Comment (hallyn) [inline resource.c]: Here and below, check for malloc failure.

File: resource.c [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+void res_destroy_blk(res_blk_t *res)
+{
+	for (int i = 0; i < res->res_count; i++) {
+		/* Some resource information are big and needed extra
+		 * Free memory allocated for that.
+		 */
+		switch (res->res_unit[i]->res_id) {
+		case RES_NET_IFSTAT:
+		case RES_MEM_INFOALL:
+			free((res->res_unit[i])->data.ptr);
```
Comment (hallyn) [inline resource.c]: Should you set .ptr = NULL after free?  This is called by the caller with no locks...

File: resource.c (hunk 3 above)
Comment (hallyn) [reply resource.c]: Note that a common way of doing this is to support a NULL 'out' meaning just return the length, so caller can do:
len = fn(request, NULL, 0);
s = malloc(len);
result = fn(request, s, len);

File: resource.c [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+
+	switch (res_id) {
+	case RES_KERN_RELEASE:
+		uname(&t);
+		strcpy(out, t.release);
+		break;
+
+	case RES_KERN_COMPILE_TIME:
+		uname(&t);
+		sscanf(t.version, "%*s%*s%*s%[^\t\n]", (char *) out);
```
Comment (hallyn) [inline resource.c]: I'm not sure whether this is some nifty use of a glibc feature or what, but the manpage for sscanf says that having more % specifiers than target pointers results in undefined behavior.

**Comment by hallyn** [comment]
Thanks.  When you're done, if you want to close this PR and create a new one, experience shows that might be the simplest way to avoid getting feedback on out of date patches.

**Comment by rahuyada** [comment]
@hallyn @brauner Closing this PR as I created a new one.

**Pull request closed without merging**